src/orion/algo/tpe.py

```diff
@@ -0,0 +1,351 @@
+# -*- coding: utf-8 -*-
+"""
+:mod:`orion.algo.tpe` -- Tree-structured Parzen Estimator Approach
+==================================================================
+
+.. module:: tpe
+   :platform: Unix
+   :synopsis: Tree-structured Parzen Estimator Approach
+
+"""
+
+import numpy
+from scipy.stats import norm
+
+from orion.algo.base import BaseAlgorithm
+from orion.core.utils.points import flatten_dims, regroup_dims
+
+
+def compute_max_ei_point(points, below_likelis, above_likelis):
+    """Compute ei among points based on their log likelihood and return the point with max ei.
+
+    :param points: list of point with real values.
+    :param below_likelis: list of log likelihood for each point in the good GMM.
+    :param above_likelis: list of log likelihood for each point in the bad GMM.
+    """
+    max_ei = -numpy.inf
+    point_index = 0
+    for i, (lik_b, lik_a) in enumerate(zip(below_likelis, above_likelis)):
+        ei = lik_b - lik_a
+        if ei > max_ei:
+            max_ei = ei
+            point_index = i
+    return points[point_index]
+
+
+# pylint:disable=assignment-from-no-return
+def adaptive_parzen_estimator(mus, low, high,
+                              prior_weight=1.0,
+                              equal_weight=False,
+                              flat_num=25):
+    """Return the sorted mus, the corresponding sigmas and weights with adaptive kernel estimator.
+
+    :param mus: list of real values for observed mus.
+    :param low: real value for lower bound of points.
+    :param high: real value for upper bound of points.
+    :param prior_weight: real value for the weight of prior point.
+    :param equal_weight: bool value indicating if all points with equal weights.
+    :param flat_num: int value indicating the number of latest points with equal weights,
+                     it is only valid if `equal_weight` is False.
+    """
+    def update_weights(total_num):
+        """Generate weights for all components"""
+        if total_num < flat_num or equal_weight:
+            return numpy.ones(total_num)
+
+        ramp_weights = numpy.linspace(1.0 / total_num, 1.0, num=total_num - flat_num)
+        flat_weights = numpy.ones(flat_num)
+        return numpy.concatenate([ramp_weights, flat_weights])
+
+    mus = numpy.asarray(mus)
+
+    prior_mu = (low + high) * 0.5
+    prior_sigma = (high - low) * 1.0
+
+    size = len(mus)
+    if size > 1:
+        order = numpy.argsort(mus)
+        sorted_mus = mus[order]
+        prior_mu_pos = numpy.searchsorted(sorted_mus, prior_mu)
+
+        weights = update_weights(size)
+
+        mixture_mus = numpy.zeros(size + 1)
+        mixture_mus[:prior_mu_pos] = sorted_mus[:prior_mu_pos]
+        mixture_mus[prior_mu_pos] = prior_mu
```

Comment:
Why do you add the mean of the dimension? Why not stick to points?

Reply:
This is more like the purpose of the prior_weight, take the dimension interval into consideration but people can decide the weight of it.

Reply:
Sorry, I still don't understand. `mixture_mus` is basically the list of points. Why add the mean to this list?